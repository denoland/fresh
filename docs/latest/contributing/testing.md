---
description: |
  Learn how to test your Fresh contributions effectively.
---

# Testing Changes

This guide covers how to test your Fresh contributions to ensure they work
correctly and don't break existing functionality.

## Required Testing Before Submitting

Before submitting any pull request, you **must** run:

```sh Terminal
deno task ok
```

This essential command runs:

- **Format check** (`deno fmt --check`)
- **Linting** (`deno lint`)
- **Type checking** (`deno task check:types`)
- **Test suite** (`deno task test`)

All of these must pass for your PR to be accepted.

## Test Suite Overview

Fresh uses Deno's built-in testing framework with the following patterns:

- **Test files**: Follow `*_test.ts` naming convention
- **Snapshot tests**: Stored in `__snapshots__/` directories
- **Parallel execution**: Tests run with `--parallel` flag for speed
- **Permissions**: All tests require `-A` (all permissions) flag

### Running Tests

```sh Terminal
# Run all tests (required before submitting)
deno task test

# Run tests for a specific file
deno test -A packages/fresh/src/app_test.ts

# Run tests with coverage
deno test -A --coverage

# Run a single test case
deno test -A --filter "test name pattern"
```

### Known Local Test Issues

Some tests may fail when run locally but will pass in CI. These can be safely
ignored:

- `Could not find server address`
- `Text file busy (os error 26)`

These are environment-specific issues that don't indicate problems with your
changes.

## Testing Your Changes

### 1. Unit and Integration Tests

Add tests for new functionality in the same directory:

```ts my_feature_test.ts
import { assertEquals } from "@std/testing/asserts";
import { myNewFeature } from "./my_feature.ts";

Deno.test("myNewFeature should work correctly", () => {
  const result = myNewFeature("input");
  assertEquals(result, "expected output");
});
```

### 2. Documentation Website Testing

The Fresh documentation website (`www/`) uses local Fresh packages, making it an
excellent integration test:

```sh Terminal
# Start the documentation website
deno task www

# Build the documentation website
deno task build-www

# Test specific documentation features
deno task test:www
```

Changes to core Fresh functionality will be reflected immediately in the
documentation site.

### 3. Vite Plugin Testing

For changes to the Vite plugin (`packages/plugin-vite/`):

```sh Terminal
# Run the Vite plugin demo
deno task demo

# Build the demo to test production builds
deno task demo:build

# Start the built demo
deno task demo:start
```

### 4. Fresh Init Testing

Test the Fresh project scaffolding tool:

```sh Terminal
# Create a test project with default settings
deno run -A packages/init/src/mod.ts test-project

# Create a project with Tailwind
deno run -A packages/init/src/mod.ts test-project --tailwind

# Create a project with legacy builder
deno run -A packages/init/src/mod.ts test-project --builder
```

## Testing Workflows by Feature Area

### Core Framework Changes

1. **Run unit tests**: `deno test -A packages/fresh/src/`
2. **Test with documentation site**: `deno task www`
3. **Test with demo**: `deno task demo`
4. **Create external test project** using
   [local development setup](./local-development)

### Routing Changes

1. **Test file-system routing**: Create test routes in `www/routes/`
2. **Test dynamic routes**: Verify parameter extraction works
3. **Test middleware**: Check middleware execution order

### Islands and Client Code

1. **Test SSR rendering**: Ensure islands render on server
2. **Test hydration**: Verify client-side functionality
3. **Test in browser**: Manual testing with developer tools
4. **Check bundle size**: Ensure no unexpected bloat

### Build System Changes

1. **Test development mode**: `deno task demo`
2. **Test production builds**: `deno task demo:build && deno task demo:start`
3. **Test with external project**: Use `links` configuration
4. **Verify generated files**: Check `_fresh/` directory contents

## Testing Best Practices

### Writing Good Tests

```ts
// ✅ Good: Descriptive name and clear expectations
Deno.test("Context should parse URL parameters correctly", () => {
  const ctx = new Context(
    new Request("http://localhost/users/123"),
    new URL("http://localhost/users/123"),
    mockInfo,
    "/users/:id",
    { id: "123" },
    config,
    () => Promise.resolve(new Response()),
    buildCache,
  );

  assertEquals(ctx.params.id, "123");
});

// ❌ Avoid: Vague names and unclear purpose
Deno.test("test context", () => {
  // unclear what this tests
});
```

### Performance Testing

```sh Terminal
# Measure test execution time
time deno task test

# Profile memory usage during tests
deno test -A --inspect --inspect-wait packages/fresh/src/app_test.ts
```

### Testing Error Conditions

```ts
Deno.test("should handle missing route gracefully", async () => {
  const response = await handler(new Request("http://localhost/nonexistent"));
  assertEquals(response.status, 404);
});

Deno.test("should throw HttpError for invalid input", () => {
  assertThrows(
    () => validateInput(""),
    HttpError,
    "Input cannot be empty",
  );
});
```

## Snapshot Testing

Fresh uses snapshot testing for complex output validation:

```ts
import { assertSnapshot } from "@std/testing/snapshot";

Deno.test("should render component correctly", async (t) => {
  const html = await renderComponent();
  await assertSnapshot(t, html);
});
```

To update snapshots when output intentionally changes:

```sh Terminal
deno test -A --update-snapshots packages/fresh/src/render_test.ts
```

## Manual Testing Checklist

Before submitting your PR, manually verify:

- [ ] **Development server starts** without errors
- [ ] **Hot reload works** for your changes
- [ ] **Production build completes** successfully
- [ ] **Generated bundles are reasonable** in size
- [ ] **Browser developer tools** show no console errors
- [ ] **Existing functionality** still works as expected

## CI/CD Testing

Your changes will be automatically tested in GitHub Actions CI when you submit a
PR. The CI runs:

- Tests on multiple Deno versions
- Tests on different operating systems (Linux, macOS, Windows)
- Linting and formatting checks
- Type checking across the entire codebase

## Debugging Test Failures

### Common Issues

**Permission Errors**

```sh Terminal
# Ensure you're using -A flag
deno test -A path/to/test.ts
```

**Import Resolution**

```sh Terminal
# Clear Deno cache if modules seem stale
deno cache --reload path/to/test.ts
```

**Snapshot Mismatches**

```sh Terminal
# Update snapshots if output intentionally changed
deno test -A --update-snapshots path/to/test.ts
```

### Debug Mode

```sh Terminal
# Run tests with debugger
deno test -A --inspect-brk packages/fresh/src/app_test.ts

# Enable verbose logging
FRESH_DEBUG=true deno test -A path/to/test.ts
```

## Getting Help

If you encounter issues while testing:

1. **Check existing issues**:
   [GitHub Issues](https://github.com/denoland/fresh/issues)
2. **Review test patterns**: Look at similar tests in the codebase
3. **Join the community**: [Deno Discord server](https://discord.gg/deno)

Remember: thorough testing helps ensure Fresh remains stable and reliable for
all users! 🧪
