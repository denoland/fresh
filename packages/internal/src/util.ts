export * from "../../test-utils/src/util.ts";
