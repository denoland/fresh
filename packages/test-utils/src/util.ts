export async function waitFor(
  fn: () => Promise<unknown> | unknown,
): Promise<void> {
  let now = Date.now();
  const limit = now + 2000;

  while (now < limit) {
    try {
      if (await fn()) return;
    } catch (err) {
      if (now > limit) {
        throw err;
      }
    } finally {
      await new Promise((r) => setTimeout(r, 250));
      now = Date.now();
    }
  }

  throw new Error(`Timed out`);
}

export function usingEnv(name: string, value: string): Disposable {
  const prev = Deno.env.get(name);
  Deno.env.set(name, value);
  return {
    [Symbol.dispose]: () => {
      if (prev === undefined) {
        Deno.env.delete(name);
      } else {
        Deno.env.set(name, prev);
      }
    },
  };
}
