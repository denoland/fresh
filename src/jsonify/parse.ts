import {
  HOLE,
  INFINITY_NEG,
  INFINITY_POS,
  NAN,
  NULL,
  UNDEFINED,
  ZERO_NEG,
} from "./constants.ts";

// deno-lint-ignore no-explicit-any
export type CustomParser = Record<string, (value: any) => unknown>;

export function parse<T = unknown>(
  value: string,
  custom?: CustomParser | undefined,
): T {
  const data = JSON.parse(value);

  const hydrated = new Array(data.length);
  // deno-lint-ignore no-explicit-any
  unpack(data, hydrated, 0, custom) as any;
  return hydrated[0];
}

function unpack(
  arr: unknown[],
  hydrated: unknown[],
  idx: number,
  custom: CustomParser | undefined,
): void {
  if (idx in hydrated) return;

  const current = arr[idx];
  if (typeof current === "number") {
    switch (current) {
      case UNDEFINED:
        hydrated[idx] = undefined;
        return;
      case NULL:
        hydrated[idx] = null;
        return;
      case NAN:
        hydrated[idx] = NaN;
        return;
      case INFINITY_POS:
        hydrated[idx] = Infinity;
        return;
      case INFINITY_NEG:
        hydrated[idx] = -Infinity;
        return;
      case ZERO_NEG:
        hydrated[idx] = -0;
        return;
      default:
        hydrated[idx] = current;
        return;
    }
  } else if (
    typeof current === "string" || typeof current === "boolean" ||
    current === null
  ) {
    hydrated[idx] = current;
    return;
  } else if (Array.isArray(current)) {
    if (current.length > 0 && typeof current[0] === "string") {
      const name = current[0];
      if (custom !== undefined && name in custom) {
        const fn = custom[name];
        const ref = current[1];
        unpack(arr, hydrated, ref, custom);
        const value = hydrated[ref];
        hydrated[idx] = fn(value);
        return;
      }
      switch (name) {
        case "BigInt":
          hydrated[idx] = BigInt(current[1]);
          return;
        case "Date":
          hydrated[idx] = new Date(current[1]);
          return;
        case "RegExp":
          hydrated[idx] = new RegExp(current[1], current[2]);
          return;
        case "Set": {
          const set = new Set();
          for (let i = 0; i < current[1].length; i++) {
            const ref = current[1][i];
            unpack(arr, hydrated, ref, custom);
            set.add(hydrated[ref]);
          }
          hydrated[idx] = set;
          return;
        }
        case "Map": {
          const set = new Map();
          for (let i = 0; i < current[1].length; i++) {
            const refKey = current[1][i++];
            unpack(arr, hydrated, refKey, custom);
            const refValue = current[1][i];
            unpack(arr, hydrated, refValue, custom);

            set.set(hydrated[refKey], hydrated[refValue]);
          }
          hydrated[idx] = set;
          return;
        }
        case "Uint8Array":
          hydrated[idx] = b64decode(current[1]);
          return;
      }
    } else {
      const actual = new Array(current.length);
      hydrated[idx] = actual;
      for (let i = 0; i < current.length; i++) {
        const ref = current[i];
        if (ref < 0) {
          switch (ref) {
            case UNDEFINED:
              actual[i] = undefined;
              break;
            case NULL:
              actual[i] = null;
              break;
            case NAN:
              actual[i] = NaN;
              break;
            case INFINITY_POS:
              actual[i] = Infinity;
              break;
            case INFINITY_NEG:
              actual[i] = -Infinity;
              break;
            case ZERO_NEG:
              actual[i] = -0;
              break;
            case HOLE:
              continue;
          }
        } else {
          unpack(arr, hydrated, ref, custom);
          actual[i] = hydrated[ref];
        }
      }
    }
  } else if (typeof current === "object") {
    const actual: Record<string, unknown> = {};
    hydrated[idx] = actual;

    const keys = Object.keys(current);
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i];
      // deno-lint-ignore no-explicit-any
      const ref = (current as any)[key];
      if (ref < 0) {
        switch (ref) {
          case UNDEFINED:
            actual[key] = undefined;
            break;
          case NULL:
            actual[key] = null;
            break;
          case NAN:
            actual[key] = NaN;
            break;
          case INFINITY_POS:
            actual[key] = Infinity;
            break;
          case INFINITY_NEG:
            actual[key] = -Infinity;
            break;
          case ZERO_NEG:
            actual[key] = -0;
            break;
        }
      } else {
        unpack(arr, hydrated, ref, custom);
        actual[key] = hydrated[ref];
      }
    }
  }
}

function b64decode(b64: string): Uint8Array {
  const binString = atob(b64);
  const size = binString.length;
  const bytes = new Uint8Array(size);
  for (let i = 0; i < size; i++) {
    bytes[i] = binString.charCodeAt(i);
  }
  return bytes;
}
