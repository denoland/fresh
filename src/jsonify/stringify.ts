import {
  INFINITY_NEG,
  INFINITY_POS,
  NAN,
  NULL,
  UNDEFINED,
  ZERO_NEG,
} from "./constants.ts";
import { HOLE } from "./constants.ts";

// deno-lint-ignore no-explicit-any
export type Stringifiers = Record<string, (value: any) => any>;

/**
 * Serializes the following:
 *
 * - `null`
 * - `undefined`
 * - `boolean`
 * - `number`
 * - `bigint`
 * - `string`
 * - `array`
 * - `object` (no prototypes)
 * - `Uint8Array`
 * - `Date`
 * - `RegExp`
 * - `Set`
 * - `Map`
 *
 * Circular references are supported and objects with the same reference are
 * serialized only once.
 */
export function stringify(data: unknown, custom?: Stringifiers): string {
  const out: string[] = [];
  const indexes = new Map<unknown, number>();
  const res = serializeInner(out, indexes, data, custom);
  if (res < 0) {
    out.push(String(res));
  }
  return `[${out.join(",")}]`;
}

function serializeInner(
  out: string[],
  indexes: Map<unknown, number>,
  value: unknown,
  custom: Stringifiers | undefined,
): number {
  const seenIdx = indexes.get(value);
  if (seenIdx !== undefined) return seenIdx;

  if (value === undefined) return UNDEFINED;
  if (value === null) return NULL;
  if (Number.isNaN(value)) return NAN;
  if (value === Infinity) return INFINITY_POS;
  if (value === -Infinity) return INFINITY_NEG;
  if (value === 0 && 1 / value < 0) return ZERO_NEG;

  const idx = out.length;
  out.push("");
  indexes.set(value, idx);

  let str = "";

  if (typeof value === "number") {
    str += String(value);
  } else if (typeof value === "boolean") {
    str += String(value);
  } else if (typeof value === "bigint") {
    str += `["BigInt","${value}"]`;
  } else if (typeof value === "string") {
    str += JSON.stringify(value);
  } else if (Array.isArray(value)) {
    str += "[";
    for (let i = 0; i < value.length; i++) {
      if (i in value) {
        str += serializeInner(out, indexes, value[i], custom);
      } else {
        str += HOLE;
      }

      if (i < value.length - 1) {
        str += ",";
      }
    }
    str += "]";
  } else if (typeof value === "object") {
    if (custom !== undefined) {
      for (const k in custom) {
        const fn = custom[k];
        if (fn === undefined) continue;

        const res = fn(value);
        if (res === undefined) continue;

        serializeInner(out, indexes, res, custom);
        str = `["${k}",${idx + 1}]`;
        out[idx] = str;
        return idx;
      }
    }

    if (value instanceof Date) {
      str += `["Date","${value.toISOString()}"]`;
    } else if (value instanceof RegExp) {
      str += `["RegExp",${JSON.stringify(value.source)}, "${value.flags}"]`;
    } else if (value instanceof Uint8Array) {
      str += `["Uint8Array","${b64encode(value)}"]`;
    } else if (value instanceof Set) {
      const items = new Array(value.size);
      let i = 0;
      value.forEach((v) => {
        items[i++] = serializeInner(out, indexes, v, custom);
      });
      str += `["Set",[${items.join(",")}]]`;
    } else if (value instanceof Map) {
      const items = new Array(value.size * 2);
      let i = 0;
      value.forEach((v, k) => {
        items[i++] = serializeInner(out, indexes, k, custom);
        items[i++] = serializeInner(out, indexes, v, custom);
      });
      str += `["Map",[${items.join(",")}]]`;
    } else {
      str += "{";
      const keys = Object.keys(value);
      for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        str += JSON.stringify(key) + ":";
        // deno-lint-ignore no-explicit-any
        str += serializeInner(out, indexes, (value as any)[key], custom);

        if (i < keys.length - 1) {
          str += ",";
        }
      }
      str += "}";
    }
  } else if (typeof value === "function") {
    throw new Error(`Serializing functions is not supported.`);
  }

  out[idx] = str;
  return idx;
}

// deno-fmt-ignore
const base64abc = [
  "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O",
  "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "a", "b", "c", "d",
  "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s",
  "t", "u", "v", "w", "x", "y", "z", "0", "1", "2", "3", "4", "5", "6", "7",
  "8", "9", "+", "/",
];

/**
 * CREDIT: https://gist.github.com/enepomnyaschih/72c423f727d395eeaa09697058238727
 * Encodes a given Uint8Array, ArrayBuffer or string into RFC4648 base64 representation
 */
export function b64encode(buffer: ArrayBuffer): string {
  const uint8 = new Uint8Array(buffer);
  let result = "",
    i;
  const l = uint8.length;
  for (i = 2; i < l; i += 3) {
    result += base64abc[uint8[i - 2] >> 2];
    result += base64abc[((uint8[i - 2] & 0x03) << 4) | (uint8[i - 1] >> 4)];
    result += base64abc[((uint8[i - 1] & 0x0f) << 2) | (uint8[i] >> 6)];
    result += base64abc[uint8[i] & 0x3f];
  }
  if (i === l + 1) {
    // 1 octet yet to write
    result += base64abc[uint8[i - 2] >> 2];
    result += base64abc[(uint8[i - 2] & 0x03) << 4];
    result += "==";
  }
  if (i === l) {
    // 2 octets yet to write
    result += base64abc[uint8[i - 2] >> 2];
    result += base64abc[((uint8[i - 2] & 0x03) << 4) | (uint8[i - 1] >> 4)];
    result += base64abc[(uint8[i - 1] & 0x0f) << 2];
    result += "=";
  }
  return result;
}
