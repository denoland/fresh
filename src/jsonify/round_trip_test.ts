import { expect } from "@std/expect";
import { parse } from "./parse.ts";
import { assertSnapshot } from "@std/testing/snapshot";
import { inspect } from "node:util";
import { stringify } from "./stringify.ts";

const inner = { foo: 123 };
const references = { a: inner, b: [inner, inner] };

const circular = { a: 1, b: null as unknown };
circular.b = circular;

const TESTS = [
  undefined,
  null,
  1,
  2,
  -2,
  Infinity,
  -Infinity,
  0,
  -0,
  NaN,
  1.2 -
  1.2,
  true,
  false,
  "abc",
  1n,
  -1n,
  [1, 2, 3],
  [null, undefined, -2],
  { a: 1, b: null, c: -2 },
  new Uint8Array([1, 2, 3]),
  new Date("1990-05-31"),
  new Map([[1, null], [undefined, -2]]),
  new Set([1, 2, null, -2, NaN]),
  [1, , 3],
  /foo["]/,
  references,
  circular,
];

for (const value of TESTS) {
  Deno.test(`round trip - ${inspect(value)}`, async (t) => {
    const str = stringify(value);
    await assertSnapshot(t, str);
    const parsed = parse(str);
    expect(parsed).toEqual(value);
  });
}
