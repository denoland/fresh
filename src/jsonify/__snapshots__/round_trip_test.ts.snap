export const snapshot = {};

snapshot[`round trip - undefined 1`] = `"-1"`;

snapshot[`round trip - null 1`] = `"-2"`;

snapshot[`round trip - 1 1`] = `"[1]"`;

snapshot[`round trip - 2 1`] = `"[2]"`;

snapshot[`round trip - -2 1`] = `"[-2]"`;

snapshot[`round trip - Infinity 1`] = `"-4"`;

snapshot[`round trip - -Infinity 1`] = `"-5"`;

snapshot[`round trip - 0 1`] = `"[0]"`;

snapshot[`round trip - -0 1`] = `"-6"`;

snapshot[`round trip - NaN 1`] = `"-3"`;

snapshot[`round trip - 0 2`] = `"[0]"`;

snapshot[`round trip - true 1`] = `"[true]"`;

snapshot[`round trip - false 1`] = `"[false]"`;

snapshot[`round trip - 'abc' 1`] = `'["abc"]'`;

snapshot[`round trip - 1n 1`] = `'[["BigInt","1"]]'`;

snapshot[`round trip - -1n 1`] = `'[["BigInt","-1"]]'`;

snapshot[`round trip - [ 1, 2, 3 ] 1`] = `"[[1,2,3],1,2,3]"`;

snapshot[`round trip - [ null, undefined, -2 ] 1`] = `"[[-2,-1,1],-2]"`;

snapshot[`round trip - { a: 1, b: null, c: -2 } 1`] = `'[{"a":1,"b":-2,"c":2},1,-2]'`;

snapshot[`round trip - Uint8Array(3) [ 1, 2, 3 ] 1`] = `'[["Uint8Array","AQID"]]'`;

snapshot[`round trip - 1990-05-31T00:00:00.000Z 1`] = `'[["Date","1990-05-31T00:00:00.000Z"]]'`;

snapshot[`round trip - Map(2) { 1 => null, undefined => -2 } 1`] = `'[["Map",[1,-2,-1,2]],1,-2]'`;

snapshot[`round trip - Set(5) { 1, 2, null, -2, NaN } 1`] = `'[["Set",[1,2,-2,3,-3]],1,2,-2]'`;

snapshot[`round trip - [ 1, <1 empty item>, 3 ] 1`] = `"[[1,-7,2],1,3]"`;

snapshot[`round trip - /foo["]/ 1`] = `'[["RegExp","foo[\\\\"]", ""]]'`;

snapshot[`round trip - { a: { foo: 123 }, b: [ { foo: 123 }, { foo: 123 } ] } 1`] = `'[{"a":1,"b":3},{"foo":2},123,[1,1]]'`;

snapshot[`round trip - <ref *1> { a: 1, b: [Circular *1] } 1`] = `'[{"a":1,"b":0},1]'`;
